apps/aecore/src/aec_fork_signalling.erl

```diff
@@ -0,0 +1,350 @@
+%%%
+%%% The worst case scenario is restarting the node at height H2 - 1. If the
+%%% node has the top key block at height H2 - 1 and the results map is empty,
+%%% it needs to start a worker process the will search for a block at HE - 1,
+%%% and then it will start computing the fork result. Some components of the
+%%% system may be waiting for the final result.
+%%%
+%%% @end
+%%%============================================================================
+-module(aec_fork_signalling).
+
+-behaviour(gen_server).
+
+%% API
+-export([start_link/0,
+         get_fork_result/2,
+         compute_fork_result/2
+        ]).
+
+%% gen_server callbacks
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2
+        ]).
+
+-export([worker_process/4]).
+
+-define(SERVER, ?MODULE).
+
+-type header_hash()      :: aec_blocks:block_header_hash().
+
+-type worker()           :: {pid(), reference()}.
+
+-type result()           :: true | false | pending.
+
+%% The pointer is a key block hash which belongs to the last signalling block.
+-type pointer()          :: header_hash().
+
+-record(state, {fork     :: aec_hard_forks:fork() | undefined,
+                results  :: #{header_hash() := result() | pointer()},
+                workers  :: #{worker() := header_hash()}
+               }).
+
+-define(IS_RESULT(X), X =:= true orelse X =:= false orelse X =:= pending).
+-define(IS_POINTER(X), is_binary(X)).
+
+%% API
+
+start_link() ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+get_fork_result(Block, BlockHash) ->
+    gen_server:call(?SERVER, {get_fork_result, Block, BlockHash}).
+
+compute_fork_result(Block, BlockHash) ->
+    gen_server:cast(?SERVER, {compute_fork_result, Block, BlockHash}).
+
+%% gen_server callbacks
+
+init([]) ->
+    process_flag(trap_exit, true),
+    {ok, #state{fork = aec_hard_forks:fork(), results = maps:new(),
+                workers = maps:new()}}.
+
+handle_call({get_fork_result, Block, BlockHash}, _From, State) ->
+    {Reply, State1} = handle_get_fork_result(Block, BlockHash, State),
+    {reply, Reply, State1};
+handle_call(_Request, _From, State) ->
+    {reply, ingnored, State}.
+
+handle_cast({compute_fork_result, Block, BlockHash}, State) ->
+    {noreply, handle_compute_fork_result(Block, BlockHash, State)};
+handle_cast(_Request, State) ->
+    {noreply, State}.
+
+handle_info({worker_msg, Pid, Msg}, State) ->
+    {noreply, handle_worker_msg(Pid, Msg, State)};
+handle_info({'DOWN', Ref, process, Pid, Rsn}, State) ->
+    {noreply, handle_down({Pid, Ref}, Rsn, State)};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Rsn, #state{workers = Workers} = State) ->
+    maps:foldl(fun(W, _BH, S1) -> kill_worker(W, S1) end, State, Workers).
+
+%% Internal functions
+
+handle_get_fork_result(Block, BlockHash,
+                       #state{fork = Fork, results = Results} = State) ->
+    case is_fork_height_block(Block, Fork) of
```

Comment:
Does this mean that this is meant to be called with argument the block and block hash of the first key block that may have the new consensus protocol? I would rather expect relevant argument to be its predecessor.

If node needs to build key block candidate at fork height, it does not have the block at that height yet...

Reply:
Fixed in 4d496b383